File: backend/src/main/java/com/groove/order/entity/OrderStatus.java
package com.groove.order.entity;

import java.util.List;

/** 주문 상태. */
public enum OrderStatus {

	PENDING,
	PAID,
	PREPARING,
	SHIPPED,
	DELIVERED,
	CANCELED,
	REFUNDED;

	/** 결제가 끝나 실제 구매로 보는 상태. 통계·추천의 "PAID 이상" 기준. */
	public static final List<OrderStatus> PAID_OR_LATER = List.of(PAID, PREPARING, SHIPPED, DELIVERED);

	public boolean isCancelable() {
		return this == PENDING || this == PAID;
	}

	/** 관리자 상태 전이(PATCH /admin/orders/{id}/status)에서 허용되는 전이만 true. */
	public boolean canTransitionTo(OrderStatus next) {
		return switch (this) {
			case PAID -> next == PREPARING || next == CANCELED;
			case PREPARING -> next == SHIPPED || next == CANCELED;
			case SHIPPED -> next == DELIVERED;
			default -> false;
		};
	}
}

File: frontend/src/utils/orderStatus.ts
import type { BadgeVariant } from '@/components/common/Badge';
import type { OrderStatus } from '@/types/order';

export const ORDER_STATUSES: readonly OrderStatus[] = [
  'PENDING',
  'PAID',
  'PREPARING',
  'SHIPPED',
  'DELIVERED',
  'CANCELED',
  'REFUNDED',
];

const ORDER_STATUS_SET = new Set<string>(ORDER_STATUSES);

export const ORDER_STATUS_LABEL: Record<OrderStatus, string> = {
  PENDING: '결제대기',
  PAID: '결제완료',
  PREPARING: '배송준비',
  SHIPPED: '배송중',
  DELIVERED: '배송완료',
  CANCELED: '취소',
  REFUNDED: '환불',
};

export const ORDER_STATUS_BADGE: Record<OrderStatus, BadgeVariant> = {
  PENDING: 'accent',
  PAID: 'success',
  PREPARING: 'success',
  SHIPPED: 'success',
  DELIVERED: 'neutral',
  CANCELED: 'danger',
  REFUNDED: 'danger',
};

/** 주문 상세 타임라인에 쓰는 정상 흐름 5단계. */
export const ORDER_STATUS_STEPS = ['PENDING', 'PAID', 'PREPARING', 'SHIPPED', 'DELIVERED'] as const;

const CANCELABLE_STATUSES = new Set<OrderStatus>(['PENDING', 'PAID']);

export const isCancelableStatus = (status: OrderStatus): boolean => CANCELABLE_STATUSES.has(status);

/** 관리자 상태 전이표. 나머지 상태는 전이 불가(빈 배열). */
export const ADMIN_ORDER_TRANSITIONS: Record<OrderStatus, OrderStatus[]> = {
  PENDING: [],
  PAID: ['PREPARING', 'CANCELED'],
  PREPARING: ['SHIPPED', 'CANCELED'],
  SHIPPED: ['DELIVERED'],
  DELIVERED: [],
  CANCELED: [],
  REFUNDED: [],
};

export const isOrderStatus = (value: unknown): value is OrderStatus =>
  typeof value === 'string' && ORDER_STATUS_SET.has(value);
